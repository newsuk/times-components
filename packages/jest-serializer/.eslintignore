**/node_modules/*
**/coverage/*
dist
rnw.js
!src/rnw.js
