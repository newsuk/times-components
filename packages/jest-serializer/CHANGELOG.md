# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

<a name="3.1.4"></a>
## [3.1.4](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.1.3...@times-components/jest-serializer@3.1.4) (2018-10-16)

**Note:** Version bump only for package @times-components/jest-serializer





<a name="3.1.3"></a>
## [3.1.3](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.1.2...@times-components/jest-serializer@3.1.3) (2018-10-11)

**Note:** Version bump only for package @times-components/jest-serializer





<a name="3.1.2"></a>
## [3.1.2](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.1.1...@times-components/jest-serializer@3.1.2) (2018-09-27)

**Note:** Version bump only for package @times-components/jest-serializer





<a name="3.1.1"></a>
## [3.1.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.1.0...@times-components/jest-serializer@3.1.1) (2018-08-29)

**Note:** Version bump only for package @times-components/jest-serializer





<a name="3.1.0"></a>
# [3.1.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.0.1...@times-components/jest-serializer@3.1.0) (2018-08-21)


### Features

* squash styles ([c051768](https://github.com/newsuk/times-components/commit/c051768))





<a name="3.0.1"></a>
## [3.0.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@3.0.0...@times-components/jest-serializer@3.0.1) (2018-08-09)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="3.0.0"></a>
# [3.0.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.1.1...@times-components/jest-serializer@3.0.0) (2018-08-09)


### Features

* split out packages and clean up deps ([#1183](https://github.com/newsuk/times-components/issues/1183)) ([85aca6b](https://github.com/newsuk/times-components/commit/85aca6b))


### BREAKING CHANGES

* split up markup into separate packages




<a name="2.1.1"></a>
## [2.1.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.1.0...@times-components/jest-serializer@2.1.1) (2018-08-07)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.1.0"></a>
# [2.1.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.6...@times-components/jest-serializer@2.1.0) (2018-08-07)


### Features

* add depcheck and cleanup ([#1169](https://github.com/newsuk/times-components/issues/1169)) ([b5c058c](https://github.com/newsuk/times-components/commit/b5c058c))




<a name="2.0.6"></a>
## [2.0.6](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.5...@times-components/jest-serializer@2.0.6) (2018-08-02)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.5"></a>
## [2.0.5](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.4...@times-components/jest-serializer@2.0.5) (2018-07-31)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.4"></a>
## [2.0.4](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.3...@times-components/jest-serializer@2.0.4) (2018-07-31)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.3"></a>
## [2.0.3](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.2...@times-components/jest-serializer@2.0.3) (2018-07-26)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.2"></a>
## [2.0.2](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.1...@times-components/jest-serializer@2.0.2) (2018-07-23)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.1"></a>
## [2.0.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@2.0.0...@times-components/jest-serializer@2.0.1) (2018-07-20)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="2.0.0"></a>
# 2.0.0 (2018-07-18)


### Bug Fixes

* dextrose ([#919](https://github.com/newsuk/times-components/issues/919)) ([a8fa1d4](https://github.com/newsuk/times-components/commit/a8fa1d4))
* linting ([#878](https://github.com/newsuk/times-components/issues/878)) ([9df1abf](https://github.com/newsuk/times-components/commit/9df1abf))


### Features

* make rnw bundles ([#958](https://github.com/newsuk/times-components/issues/958)) ([c412a57](https://github.com/newsuk/times-components/commit/c412a57))
* omit function prop values from snapshots ([#994](https://github.com/newsuk/times-components/issues/994)) ([6050e4e](https://github.com/newsuk/times-components/commit/6050e4e))
* paidOnly videos and new video package ([#889](https://github.com/newsuk/times-components/issues/889)) ([2e57f4b](https://github.com/newsuk/times-components/commit/2e57f4b))


* breaking change: compose serializers (#1029) ([2e45d57](https://github.com/newsuk/times-components/commit/2e45d57)), closes [#1029](https://github.com/newsuk/times-components/issues/1029)


### BREAKING CHANGES

* add options object to jest-configurator
* feat: hoist dev specific rnw styles in snapshots a la styled components
* feat: remove empty values from snapshots e.g. stickyHeaderIndices
* feat: web retains className even if empty and native does not
* chore: remove coverage flags from test scripts
* chore: bump enzyme-to-json




<a name="1.0.9"></a>
## [1.0.9](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.8...@times-components/jest-serializer@1.0.9) (2018-07-11)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.8"></a>
## [1.0.8](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.7...@times-components/jest-serializer@1.0.8) (2018-07-10)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.7"></a>
## [1.0.7](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.6...@times-components/jest-serializer@1.0.7) (2018-07-09)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.6"></a>
## [1.0.6](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.5...@times-components/jest-serializer@1.0.6) (2018-07-07)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.5"></a>
## [1.0.5](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.4...@times-components/jest-serializer@1.0.5) (2018-07-06)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.4"></a>
## [1.0.4](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.3...@times-components/jest-serializer@1.0.4) (2018-07-02)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.3"></a>
## [1.0.3](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.2...@times-components/jest-serializer@1.0.3) (2018-06-27)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.2"></a>
## [1.0.2](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.1...@times-components/jest-serializer@1.0.2) (2018-06-22)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.1"></a>
## [1.0.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@1.0.0...@times-components/jest-serializer@1.0.1) (2018-06-20)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="1.0.0"></a>
# [1.0.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.3.1...@times-components/jest-serializer@1.0.0) (2018-06-15)


* breaking change: compose serializers (#1029) ([2e45d57](https://github.com/newsuk/times-components/commit/2e45d57)), closes [#1029](https://github.com/newsuk/times-components/issues/1029)


### BREAKING CHANGES

* add options object to jest-configurator
* feat: hoist dev specific rnw styles in snapshots a la styled components
* feat: remove empty values from snapshots e.g. stickyHeaderIndices
* feat: web retains className even if empty and native does not
* chore: remove coverage flags from test scripts
* chore: bump enzyme-to-json




<a name="0.3.1"></a>
## [0.3.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.3.0...@times-components/jest-serializer@0.3.1) (2018-06-13)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="0.3.0"></a>
# [0.3.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.2.1...@times-components/jest-serializer@0.3.0) (2018-06-07)


### Features

* omit function prop values from snapshots ([#994](https://github.com/newsuk/times-components/issues/994)) ([6050e4e](https://github.com/newsuk/times-components/commit/6050e4e))




<a name="0.2.1"></a>
## [0.2.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.2.0...@times-components/jest-serializer@0.2.1) (2018-05-31)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="0.2.0"></a>
# [0.2.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.1.4...@times-components/jest-serializer@0.2.0) (2018-05-29)


### Features

* make rnw bundles ([#958](https://github.com/newsuk/times-components/issues/958)) ([c412a57](https://github.com/newsuk/times-components/commit/c412a57))




<a name="0.1.4"></a>
## [0.1.4](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.1.3...@times-components/jest-serializer@0.1.4) (2018-05-25)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="0.1.3"></a>
## [0.1.3](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.1.2...@times-components/jest-serializer@0.1.3) (2018-05-15)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="0.1.2"></a>
## [0.1.2](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.1.1...@times-components/jest-serializer@0.1.2) (2018-05-04)


### Bug Fixes

* dextrose ([#919](https://github.com/newsuk/times-components/issues/919)) ([a8fa1d4](https://github.com/newsuk/times-components/commit/a8fa1d4))




<a name="0.1.1"></a>
## [0.1.1](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.1.0...@times-components/jest-serializer@0.1.1) (2018-04-25)




**Note:** Version bump only for package @times-components/jest-serializer

<a name="0.1.0"></a>
# [0.1.0](https://github.com/newsuk/times-components/compare/@times-components/jest-serializer@0.0.2...@times-components/jest-serializer@0.1.0) (2018-04-23)


### Features

* paidOnly videos and new video package ([#889](https://github.com/newsuk/times-components/issues/889)) ([2e57f4b](https://github.com/newsuk/times-components/commit/2e57f4b))




<a name="0.0.2"></a>
## 0.0.2 (2018-04-17)


### Bug Fixes

* linting ([#878](https://github.com/newsuk/times-components/issues/878)) ([9df1abf](https://github.com/newsuk/times-components/commit/9df1abf))
