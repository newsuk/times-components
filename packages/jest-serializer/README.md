# jest-serializer

A suite of tools to enable you to "curate" the "perfect" Jest snapshot. See
[this blog post](https://medium.com/@craigbilner/expect-anything-tomatchsnapshot-here-be-dragons-26516e02de8c)
for more details on why this package was developed. The primary focus is on JSX
in snapshots rather than other data structures.

It's designed to aid you with your tests that use Jest with no other
requirements \* other than importing the desired serializers for your needs.

\* there's currently a dependency on `react-native-web` that needs removal.

### Supported

- Jest@21.2.1
- Node@8.11.3

## Getting started

```
yarn add @times-components/jest-serializer --dev
```

within your tests you can now use code that looks like this:

```javascript
import {
  addSerializers,
  enzymeRenderedSerializer,
  minimalNative
} from "@times-components/jest-serializer";

addSerializers(expect, enzymeRenderedSerializer(), minimalNative);
```

and any snapshot tests that run in the context of this serializer will generate
snapshots based on these
[custom serializers](http://jestjs.io/docs/en/configuration.html#snapshotserializers-array-string).
The context being the "test file" currently under test.

## Serializers

**addSerializers(expect: JestExpect, ...serializers: Serializer[]): void**

Implicitly add your custom Jest serializer(s) by using `addSerializers`. This
takes Jest's global `expect` followed by any serializers you would like to use
e.g.

```javascript
addSerializers(expect, serializer1, serializer2);
```

These are the serializers that are currently provided:

### compose

**compose(printer: Printer, ...transformers: Transformer[]): Serializer**

A special type of serializer that takes a `printer` followed by the transformers
you would like to apply.

A transformer is any of the serializers below but with "Transform" appended.
This allows us to only visit each node once and perform each transformation
(from left to right) on it. Accordingly, order is important for transformers
that have side-effects that another transformer was relying on e.g.

```javascript
import { AppRegistry } from "react-native-web";

compose(
  stylePrinter,
  minimalWebTransform,
  rnwTransform(AppRegistry, ["color"])
);
```

is the same as running the `minimalWeb` and `rnw` serializers together, and if
`rnw` had a dependency on what `minimalWeb` was removing, you would need to swap
them round.

### Enzyme Serializers

When using enzyme in conjunction with
[enzyme-to-json](https://github.com/adriantoine/enzyme-to-json) you can use
[mount](https://github.com/airbnb/enzyme#full-dom-rendering) to give you access
to
[these modes](https://github.com/adriantoine/enzyme-to-json/blob/master/docs/modes.md).
They've been wrapped in the serializers below for convenience and clarity

#### enzymeRenderedSerializer

**enzymeRenderedSerializer(): Serializer**

Gives you your mounted Enzyme component tree as if it was rendered. i.e.
`<div>`s and `<span>`s. This useful if you don't care to manipulate the tree in
anyway.

#### enzymeRootSerializer

**enzymeRootSerializer(): Serializer**

While you may have mounted the component you can still ask to just serialize the
root component only. May be useful if you have a mixture of snapshot
requirements involving both parent and children. However you may use Enzyme's
own API to pick the correct thing to snapshot to begin with.

#### enzymeTreeSerializer

**enzymeTreeSerializer(): Serializer**

The majority of the time you may be using the rendered version but this one
allows you to see the "host" components and therefore allow you to refine the
snapshot in various ways based on them. Can be used in conjunction with the
`replace` serializer to keep the snapshots focused.

e.g.

```jsx harmony
<ParentComponent>
  <div>
    <ChildComponent>
      <span>Hello world!</span>
    </ChildComponent>
  </div>
</ParentComponent>
```

### flattenStyle

**flattenStyle: Serializer**

If you're using Enzyme you may find that styles (depending on how they've been
written) may appear as a number (a pointer to a RN stylesheet style) or an Array
(a composition of styles). This squashes it all for you, and allows you to see
the styles as intended.

On web you may want to also use `hoistStyle` to clean up the snapshot noise
further.

### hoistStyle

**hoistStyle: Serializer**

While on Native it may make sense to show a snapshot with inline styles on web,
it's perhaps a little more idiomatic to see it as a `className` and the style to
reside in a `style` block. This is a better reflection of how it'll be rendered
on the platform (RNW magic) and keeps the snapshots a little leaner too.

For the styles to be shown, use with `compose` and the `stylePrinter`.

### minimalise

**minimalise((value: Object, key: string): boolean): Serializer**

Takes a function that allows you to omit certain properties that you're not
interested in seeing. Often snapshots will spew out many `props` that are
irrelevant to your tests (especially on RN), this serializer is a key tool in
keeping them focused e.g.

```javascript
addSerializers(expect, minimalise((value, key) => key === "style"));
```

will strip all styles from your JSX nodes.

### minimalNative

**minimalNative: Serializer**

A convenience serializer to remove the plethora of `prop`s RN prints by default
that may change between RN version but have no baring on your tests. In
particular it removes `[Function]` values because they should be tested with
either interactive tests and/or a type system rather than with snapshots.

Refer to the code for the latest removed values.

### minimalRnw

**minimalRnw(AppRegistry): Serializer**

An example of perhaps where this package is likely to move to as patterns and
common use cases appear. This combines `minimalWeb` with `rnw` much like a babel
preset.

### minimalWeb

**minimalWeb: Serializer**

A convenience serializer to remove the plethora of `prop`s React prints by
default that may change between React and/or react-native-web versions but have
no baring on your tests. In particular it removes `[Function]` values because
they should be tested with either interactive tests and/or a type system rather
than with snapshots.

Refer to the code for the latest removed values.

### replace

**type Replacer = (node: ReactNode, props: ReactProps, children: ReactChildren):
{ node?: ReactNode, props?: ReactProps, children?: ReactChildren };**

**replace({ [ComponentName: string]: Replacer }): Serializer**

A serializer that may be best suited to `jest.mock` but in some cases you may
find it desirable to simply switch out bits of your tree for brevity and focus
independent of what `jest.mock` provides or because you don't want to globally
mock the dependency. `jest-mock` also doesn't play well with web or named member
imports.

### replaceProp

**replaceProp((value: Object, key: string): Object): Serializer**

Similar to the `minimalise` serializer, rather than omitting values you can
replace them. This is handy for occasions when a prop value has a very long
completely unreadable value that you would like to "monitor" and see on a PR
that a change has been made but at the same time not destroy the snapshots
readability.

### rnw

**rnw(AppRegistry, string[]): Serializer**

If you're using [react-native-web](https://github.com/necolas/react-native-web)
and Jest snapshots on the web platform to ensure your components are rendering
the expected web output, there'll be a lot of `className` noise.

Use this serializer with any style properties you're interested in, and when
`compose`d with the `stylePrinter` you can see these printed in a `style` block
e.g.

```javascript
import { AppRegistry } from "react-native-web";

compose(
  stylePrinter,
  rnwTransform(AppRegistry, ["color", "fontSize"])
);
```

will assign a generic `className` to the Node and print the styles with that
`className`, but only for `color` and `fontSize` due to RNW producing many stock
styles.

`AppRegistry` needs to come from the consumer where the styles have been cached.

## Printers

Currently there are two types of "printers". A printer has the type:

**(serialize: ((obj: Object): string), accum: {[key: string]: string}, element:
ReactNode): string**

The custom serializers can build up something they'd like to "accumulate" over
the whole snapshot such as styles, and the printer decides how to print them.

### print

Simply takes the given `Node` and uses Jest's standard `serialize` method;

### stylePrinter

A custom printer that works with `hoistStyle` and `rnw` which takes their
accumulated "bag" of styles and prints them in a `style` block

## Patterns and Practices

As discussed in the blog post at the start of the document, there are currently
numerous problems around "building trees" of React components and rendering them
out nicely on both Native and Web platforms. Enzyme doesn't have the React
Native adapter so using React Test Renderer tends to be your best bet. This
doesn't give you "host" objects for the `replace` serializer leaving you with
`jest.mock`. It's currently advisable to mock all your dependencies even if you
have the best intentions with "integration" testing. They usually lead to a
multitude of snapshot updates you don't care about which dulls your interest in
reviewing them at the PR level with little benefit.

For web using Enzyme with `mount` and a mixture of the
`enzymeRenderedSerializer` and `enzymeTreeSerializer` should give you everything
you need to start with. As with the curation of any snapshot, it's best to start
with the raw output, then prune it iteratively to make sure you both don't miss
something and/or can make informed decision on where you want to split your
snapshots.

It's particularly recommended to split out styling from rendering in tests due
to the same "too many updates" problem. It's much more likely that the snapshots
will be valued if they only update when they're supposed to and changing
`padding` shouldn't change the snapshot test for "renders capital letters". It
also highlights on a PR when a styling change is made and the style snapshot _is
not_ updated to the reviewer. This would indicate it's not being correctly
captured which can get lost with large "big picture" renders.

Another pattern you want to look out for is repeated props throughout a snapshot
which probably means they should be removed and one test should represent them.
It's discretionary but probably means the snapshots aren't focused on the tests
enough.

On Native you're likely going to want to use `minimalNative`, `flattenStyle` and
`minimalise` to focus the given tests.

On web you're similarly going to be looking at `minimalWeb`, `hoistStyle`,
`minimalise`, and `rnw`.

BE AWARE! Under the hood of `addSerializer` it's using
`expect.addSnapshotSerializer`. There's no `expect.removeSnapshotSerializer` and
they don't compose particularly well due to the various side-effects after
they've been applied. While it is possible to get a coincidental balance of
serializers at the test level (order very much dependent), try and run your
desired serializers per test file for sanity.
