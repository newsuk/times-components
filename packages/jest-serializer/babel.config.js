module.exports = api => {
  api.cache(true);
  return {
    plugins: ["@babel/plugin-proposal-object-rest-spread"],
    presets: ["module:metro-react-native-babel-preset"]
  };
};
