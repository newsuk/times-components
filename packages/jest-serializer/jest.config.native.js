module.exports = {
  collectCoverageFrom: [
    "**/jest-serializer/src/**",
    "!**/jest-serializer/src/rnw.js"
  ],
  coverageDirectory: `packages/jest-serializer/coverage/ios`,
  haste: {
    defaultPlatform: "ios",
    moduleFileExtensions: [`ios.js`, "native.js", "js", "json"],
    platforms: ["ios"],
    providesModuleNodeModules: ["react", "react-native"]
  },
  moduleNameMapper: {
    "\\.(png)$": "identity-obj-proxy"
  },
  preset: "react-native",
  rootDir: "../../",
  testMatch: [`**/jest-serializer/__tests__/ios/**.test.js`],
  transformIgnorePatterns: [
    "node_modules/(?!(react-native|react-native-linear-gradient|react-native-iphone-x-helper|@times-components)/)"
  ]
};
