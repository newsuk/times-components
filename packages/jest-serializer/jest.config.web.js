const path = require("path");

module.exports = {
  collectCoverageFrom: ["**/jest-serializer/src/**"],
  coverageDirectory: `packages/jest-serializer/coverage/web`,
  moduleFileExtensions: ["web.js", "js", "json"],
  moduleNameMapper: {
    "\\.(png)$": "identity-obj-proxy",
    "^react-native$": "react-native-web"
  },
  preset: "react-native",
  rootDir: "../../",
  setupFiles: [path.resolve(__dirname, "./setup.js")],
  testEnvironment: "jsdom",
  testMatch: [`**/jest-serializer/__tests__/web/**.test.js`],
  testURL: "http://localhost",
  transformIgnorePatterns: [
    "node_modules/(?!(react-native|react-native-linear-gradient|react-native-iphone-x-helper|@times-components)/)"
  ]
};
