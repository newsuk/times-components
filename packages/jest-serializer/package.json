{
  "name": "@times-components/jest-serializer",
  "version": "3.1.13",
  "description": "serializer",
  "main": "dist/index",
  "dev": "src/jest-serializer",
  "scripts": {
    "fmt": "eslint . --fix && prettier --write '**/*.*'",
    "prettier:diff": "prettier --list-different '**/*.*'",
    "depcheck": "depcheck --ignores='babel-*,depcheck,eslint,identity-obj-proxy,jest,prettier,react-art,webpack*' --ignore-bin-package=false --skip-missing",
    "lint": "eslint . && yarn prettier:diff && yarn depcheck",
    "test:ios": "jest --config='./jest.config.native.js'",
    "test:web": "jest --config='./jest.config.web.js'",
    "prepublishOnly": "yarn transpile",
    "cleanup-dist": "rm -rf dist",
    "transpile": "yarn cleanup-dist && babel src -d dist"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/newsuk/times-components.git"
  },
  "keywords": [
    "react-native-web",
    "react",
    "native",
    "web",
    "jest-serializer",
    "component"
  ],
  "license": "BSD-3-Clause",
  "bugs": {
    "url": "https://github.com/newsuk/times-components/issues"
  },
  "homepage": "https://github.com/newsuk/times-components#readme",
  "devDependencies": {
    "@times-components/eslint-config-thetimes": "0.8.9",
    "depcheck": "0.6.9",
    "enzyme": "3.6.0",
    "enzyme-adapter-react-16": "1.5.0",
    "eslint": "5.9.0",
    "identity-obj-proxy": "3.0.0",
    "jest": "23.3.0",
    "prettier": "1.14.3",
    "prop-types": "15.6.2",
    "react-art": "16.5.2",
    "react-native-web": "0.9.0",
    "react-test-renderer": "16.5.2"
  },
  "dependencies": {
    "css": "^2.2.1",
    "enzyme-to-json": "3.3.4",
    "hyphenate-style-name": "1.0.2",
    "lodash.isequal": "4.4.0",
    "lodash.omitby": "4.6.0"
  },
  "peerDependencies": {
    "react": ">=16",
    "react-dom": ">=16",
    "react-native": ">=0.53.0",
    "react-native-web": "0.9.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
