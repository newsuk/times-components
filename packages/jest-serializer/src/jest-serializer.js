import React from "react";
import { StyleSheet } from "react-native-web";
import omit from "lodash.omit";
import omitBy from "lodash.omitby";

function withoutProps(node) {
  return { ...node, props: {} };
}

function getType({ type }) {
  if (type instanceof Function) {
    return type.name.toLowerCase();
  }
  return type.toLowerCase();
}

function cleanSvgProps(node, props) {
  const type = getType(node);
  return type === "path" || type === "polygon" || type === "svg"
    ? omit(props, ["d", "path", "points", "viewBox"])
    : props;
}

function cleanClassNames(names) {
  const className = (names || "")
    .replace(/rn-[^-^ ]+-[^-^ ]+/g, "")
    .replace(/[ ]+/g, " ")
    .trim();

  return className.length ? { className } : {};
}

function transformRenderProps(propsToTransform, transformer) {
  return Object.entries(propsToTransform)
    .filter(([k, p]) => k !== "children" && React.isValidElement(p))
    .reduce(
      (transformed, [key, element]) => ({
        ...transformed,
        [key]: transformer(element)
      }),
      {}
    );
}

function transform(node) {
  if (!node || !node.props) return node;

  const { className, style: styles, ...props } = { ...node.props };

  const children = []
    .concat(node.children || node.props.children)
    .map(transform);

  const flattened = StyleSheet.flatten(styles);
  const style = Object.keys(flattened || {}).length ? { style: flattened } : {};

  return React.cloneElement(
    withoutProps(node),
    omitBy(
      {
        ...cleanSvgProps(node, props),
        ...transformRenderProps(props, transform),
        ...cleanClassNames(className),
        ...style
      },
      x => x === undefined
    ),
    ...children
  );
}

function test(value) {
  return !!value && value.$$typeof === Symbol.for("react.test.json");
}

function print(node, serialize) {
  return serialize(transform(node));
}

module.exports = { test, print };
