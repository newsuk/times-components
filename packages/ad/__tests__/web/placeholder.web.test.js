import React from "react";
import renderer from "react-test-renderer";
import { addSerializers, rnw } from "@times-components/jest-serializer";

import AdPlaceholder from "../../src/ad-placeholder";

jest.mock("@times-components/watermark", () => () => "watermark");

describe("AdPlaceholder", () => {
  addSerializers(expect, rnw(["height", "width"]));

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders a tiny placeholder", () => {
    const tree = renderer
      .create(<AdPlaceholder width={100} height={100} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a small placeholder", () => {
    const tree = renderer
      .create(<AdPlaceholder width={300} height={300} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a medium placeholder", () => {
    const tree = renderer
      .create(<AdPlaceholder width={728} height={300} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders a large placeholder", () => {
    const tree = renderer
      .create(<AdPlaceholder width={970} height={300} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
