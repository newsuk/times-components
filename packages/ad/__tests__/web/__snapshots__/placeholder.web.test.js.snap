// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1. Advert placeholder 1`] = `
<div>
  <div
    style={
      Object {
        "height": "300px",
        "width": "970px",
      }
    }
  >
    <div>
      <svg
        height={300}
        width={970}
      >
        <g
          fill="none"
          fillRule="evenodd"
          id="Page-1"
          opacity="0.07"
          stroke="none"
          strokeWidth="1"
        >
          <path
            fill="#1D1D1B"
            id="Shape"
          />
        </g>
      </svg>
    </div>
    <div>
      ADVERTISEMENT
    </div>
  </div>
</div>
`;
